pkg/identity/codefresh/principal.go

```diff
@@ -0,0 +1,166 @@
+	var claims struct {
+		AccountID         string `json:"account_id"`
+		AccountName       string `json:"account_name"`
+		PipelineID        string `json:"pipeline_id"`
+		PipelineName      string `json:"pipeline_name"`
+		WorkflowID        string `json:"workflow_id"`
+		Initiator         string `json:"initiator"`
+		SCMRepoURL        string `json:"scm_repo_url"`
+		SCMUsername       string `json:"scm_user_name"`
+		SCMRef            string `json:"scm_ref"`
+		SCMPullRequestRef string `json:"scm_pull_request_target_branch"`
+		RunnerEnvironment string `json:"runner_environment"`
+		PlatformURL       string `json:"platform_url"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.AccountID == "" {
+		return nil, errors.New("missing account_id in token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id in token")
+	}
+
+	if claims.WorkflowID == "" {
+		return nil, errors.New("missing workflow_id in token")
+	}
+
+	// Set default platform url in case it is missing in the token
+	if claims.PlatformURL == "" {
+		claims.PlatformURL = DefaultPlatformURL
+	}
+
+	return &workflowPrincipal{
+		subject:                    token.Subject,
+		issuer:                     token.Issuer,
+		accountID:                  claims.AccountID,
+		accountName:                claims.AccountName,
+		pipelineID:                 claims.PipelineID,
+		pipelineName:               claims.PipelineName,
+		workflowID:                 claims.WorkflowID,
+		initiator:                  claims.Initiator,
+		scmUsername:                claims.SCMUsername,
+		scmRepoURL:                 claims.SCMRepoURL,
+		scmRef:                     claims.SCMRef,
+		scmPullRequestTargetBranch: claims.SCMPullRequestRef,
+		runnerEnvironment:          claims.RunnerEnvironment,
+		platformURL:                claims.PlatformURL,
+	}, nil
+}
+
+func (w workflowPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
+
+	baseURL, err := url.Parse(w.platformURL)
+
+	if err != nil {
+		return err
+	}
+
+	// Set SAN to the <platform url>/<account name>/<pipeline name>:<account id>/<pipeline id> - for example https://g.codefresh.io/codefresh-account/oidc-test/get-token:628a80b693a15c0f9c13ab75/65e5a53e52853dc51a5b0cc1
+	// In Codefresh account names and pipeline names may be changed where as IDs do not.
+	// This pattern will give users the possibility to verify the signature using various forms of `cosign verify --certificate-identity-regexp` i.e https://g.codefresh.io/codefresh-account/oidc-test/get-token:* or https://g.codefresh.io/*:628a80b693a15c0f9c13ab75/65e5a53e52853dc51a5b0cc1
+	cert.URIs = []*url.URL{baseURL.JoinPath(w.accountName, fmt.Sprintf("%s:%s/%s", w.pipelineName, w.accountID, w.pipelineID))}
+
+	cert.ExtraExtensions, err = certificate.Extensions{
+		Issuer: w.issuer,
+		// URL of the build in Codefresh
+		BuildSignerURI:      baseURL.JoinPath("build", w.workflowID).String(),
```

Comment:
To confirm, a workflow comes from a pipeline, correct?

Typically build signer would be the same as the SAN, as it's meant to represent an stable identity rather than a more frequently changing identifier like a run ID. I think builder signer should be composed of the pipeline ID rather than workflow ID.

Reply:
A workflow is an instance of a pipeline, yes. But it Codefresh, pipeline definitions are only accessible to authenticated users - if you have a public build only the workflow run URL is accessible to unauthenticated users. So in case of an open source project that uses Codefresh, if one wanted to see the pipeline that was used to build their artifact they can only access the workflow if they are not authenticated to Codefresh. Plus the workflow contains the pipeline definition that was used at the time of building the artifact which represents the signer well.

Reply:
As for the runner_environment - it has been added recently to support Fulcio requirements. It is currently behind a feature flag - we will open it for everyone shortly.

Reply:
Thanks for the explanation, that makes sense. Would you be able to add a comment summarizing that, since this is the first example of a SAN being different than the build signer URI?

Reply:
@haydentherapper sure thing. Comment was added.